<!doctype html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>






<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.1" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="RNA-seq,Tools," />








  <link rel="shortcut icon" type="image/x-icon" href="/favicon.png?v=5.1.1" />






<meta name="description" content="MathJax.Hub.Config({   TeX: { equationNumbers: { autoNumber: &quot;AMS&quot; } } });   Abstract总结一下sashimi plot相关的方法，找一个最适合我现在状况的方法。">
<meta name="keywords" content="RNA-seq,Tools">
<meta property="og:type" content="article">
<meta property="og:title" content="How to plot sashimi plot? Tools and their usages">
<meta property="og:url" content="http://yoursite.com/2018/06/11/sashimi-plot/index.html">
<meta property="og:site_name" content="Jason&#39;s Cabin">
<meta property="og:description" content="MathJax.Hub.Config({   TeX: { equationNumbers: { autoNumber: &quot;AMS&quot; } } });   Abstract总结一下sashimi plot相关的方法，找一个最适合我现在状况的方法。">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://miso.readthedocs.io/en/fastmiso/_images/sashimi-plot-example-annotated.png">
<meta property="og:image" content="https://software.broadinstitute.org/software/igv/sites/cancerinformatics.org.igv/files/SL_Sashimi1.png">
<meta property="og:image" content="http://montgomerylab.stanford.edu/spliceplot/_images/sashimi_sample.png">
<meta property="og:image" content="http://montgomerylab.stanford.edu/spliceplot/_images/math/453a31526e1d09c7ec4c25bc12199f8824f1facb.png">
<meta property="og:updated_time" content="2019-06-14T05:54:33.015Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="How to plot sashimi plot? Tools and their usages">
<meta name="twitter:description" content="MathJax.Hub.Config({   TeX: { equationNumbers: { autoNumber: &quot;AMS&quot; } } });   Abstract总结一下sashimi plot相关的方法，找一个最适合我现在状况的方法。">
<meta name="twitter:image" content="http://miso.readthedocs.io/en/fastmiso/_images/sashimi-plot-example-annotated.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":0,"b2t":false,"scrollpercent":false},
    fancybox: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/2018/06/11/sashimi-plot/"/>





  <title>How to plot sashimi plot? Tools and their usages | Jason's Cabin</title>
</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  




<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
          m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-99090953-1', 'auto');
  ga('send', 'pageview');
</script>











  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Jason's Cabin</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">warm and cozy</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            搜索
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="搜索..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal " itemscope itemtype="http://schema.org/Article">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/06/11/sashimi-plot/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Jason Li">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Jason's Cabin">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">How to plot sashimi plot? Tools and their usages</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-06-11T22:37:53+00:00">
                2018-06-11
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/note/" itemprop="url" rel="index">
                    <span itemprop="name">note</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="page-pv"><i class="fa fa-file-o"></i>
            <span class="busuanzi-value" id="busuanzi_value_page_pv" ></span>
            </span>
          

          

          

        </div>
      </header>
    

    <div class="post-body" itemprop="articleBody">

      
      

      
        <script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

<h1 id="Abstract"><a href="#Abstract" class="headerlink" title="Abstract"></a>Abstract</h1><p>总结一下sashimi plot相关的方法，找一个最适合我现在状况的方法。</p>
<a id="more"></a>
<h1 id="What-is-sashimi-plot"><a href="#What-is-sashimi-plot" class="headerlink" title="What is sashimi_plot?"></a>What is sashimi_plot?</h1><blockquote>
<p>Sashimi plots first appeared as a part of the MISO package developed by Katz et. al. in 2010. They allow for the easy comparison of read depth, alternative splicing, and isoform structure between individual RNA-seq samples.</p>
</blockquote>
<p>This introduction comes from <a href="http://montgomerylab.stanford.edu/spliceplot/index.html" target="_blank" rel="noopener">SplicePlot</a></p>
<h1 id="现存的几种方案"><a href="#现存的几种方案" class="headerlink" title="现存的几种方案"></a>现存的几种方案</h1><h2 id="MISO"><a href="#MISO" class="headerlink" title="MISO"></a><a href="http://miso.readthedocs.io/en/fastmiso/sashimi.html" target="_blank" rel="noopener">MISO</a></h2><h3 id="Installation-amp-Dependencies"><a href="#Installation-amp-Dependencies" class="headerlink" title="Installation &amp; Dependencies"></a>Installation &amp; Dependencies</h3><p><a href="http://genes.mit.edu/burgelab/miso/software.html" target="_blank" rel="noopener">MISO whole package</a>,<br>which could be installed using <code>conda</code><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -c bioconda misopy</span><br></pre></td></tr></table></figure></p>
<p><code>sashimi_plot</code> itself only requires Python package matplotlib(version 1.1.0 or higher) as well as samtools</p>
<h3 id="Usage-amp-Materials-needed"><a href="#Usage-amp-Materials-needed" class="headerlink" title="Usage &amp; Materials needed"></a>Usage &amp; Materials needed</h3><ul>
<li>GFF file recording the informations of your alternative events</li>
<li>setting.txt<ul>
<li>bam_files</li>
<li>miso_files</li>
<li>plot settings</li>
<li>Bayes factors thresholds for –plot-bf-dist</li>
</ul>
</li>
</ul>
<p>Inputs require a gff file of the alternative events, which has to be <strong>indexed</strong> in order to be used with MISO with the <code>index_gff</code> script.</p>
<p>plot a event by running following command line from within the <code>sashimi_plot</code> directory in MISO:<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sashimi_plot --plot-event &quot;chr17:45816186:45816265:-@chr17:45815912:45815950:-@chr17:45814875:45814965:-&quot; test-data/event-data/ settings/sashimi_plot_settings.txt --output-dir test-plot</span><br></pre></td></tr></table></figure></p>
<p>the event name is listed in the GFF file</p>
<p>The <a href="(https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4542614/">paper</a>) for stand-alone sashimi plots was published in 2015.</p>
<h3 id="Output-example"><a href="#Output-example" class="headerlink" title="Output example"></a>Output example</h3><p><img src="http://miso.readthedocs.io/en/fastmiso/_images/sashimi-plot-example-annotated.png" alt="MISO sashimi_plot"><br>Noted that the junction connections were quantified and displayed by junction reads number.</p>
<h3 id="Comments"><a href="#Comments" class="headerlink" title="Comments"></a>Comments</h3><p>Since the inputs require MISO output, it means that<br>I have to run the whole packages all over. Cannot just plot from bam files.</p>
<p><a href="http://miso-users.mit.narkive.com/KCWQVe7I/sashimi-plot-without-miso" target="_blank" rel="noopener">update</a>:</p>
<blockquote>
<p>Not specifying the miso_files or specifying any directory that does not contain MISO files (or a value like “None”) should work.</p>
</blockquote>
<h2 id="IGV"><a href="#IGV" class="headerlink" title="IGV"></a><a href="https://software.broadinstitute.org/software/igv/Sashimi" target="_blank" rel="noopener">IGV</a></h2><h3 id="Installation-amp-Dependencies-1"><a href="#Installation-amp-Dependencies-1" class="headerlink" title="Installation &amp; Dependencies"></a>Installation &amp; Dependencies</h3><p>Download the software <a href="https://software.broadinstitute.org/software/igv/download" target="_blank" rel="noopener">here</a>.<br>Need <code>igvtools</code> or <a href="http://www.htslib.org/" target="_blank" rel="noopener"><code>samtools</code></a> to generate index files for each bam.</p>
<h3 id="Usage-amp-Materials-needed-1"><a href="#Usage-amp-Materials-needed-1" class="headerlink" title="Usage &amp; Materials needed"></a>Usage &amp; Materials needed</h3><p>软件可以直接操作看，也可以使用batch script控制<br>控制命令参考<a href="https://software.broadinstitute.org/software/igv/PortCommands" target="_blank" rel="noopener">Controlling IGV through a Port</a><br>但关于sashimi plot的操作似乎还是通过MISO的python 脚本来实现的</p>
<p><a href="https://groups.google.com/forum/#!topic/igv-help/bafUoKYzEVE" target="_blank" rel="noopener">igv-help of google group</a></p>
<blockquote>
<p>For scripting I recommend you use the original sashimi plot:  <a href="https://miso.readthedocs.io/en/fastmiso/sashimi.html" target="_blank" rel="noopener">https://miso.readthedocs.io/en/fastmiso/sashimi.html</a></p>
</blockquote>
<p><a href="https://software.broadinstitute.org/software/igv/Sashimi" target="_blank" rel="noopener">igv-documentation for Sashimi plot</a></p>
<blockquote>
<p>Static images of Sashimi plots can also be generated outside IGV with sashimi_plot, a Python tool which is part of the <a href="http://genes.mit.edu/burgelab/miso/" target="_blank" rel="noopener">MISO</a> package. Read more about sashimi_plot <a href="http://miso.readthedocs.io/en/fastmiso/sashimi.html" target="_blank" rel="noopener">here</a>.</p>
</blockquote>
<h3 id="Output-example-1"><a href="#Output-example-1" class="headerlink" title="Output example"></a>Output example</h3><p><img src="https://software.broadinstitute.org/software/igv/sites/cancerinformatics.org.igv/files/SL_Sashimi1.png" alt="IGV sashimi_plot"></p>
<h3 id="Comments-1"><a href="#Comments-1" class="headerlink" title="Comments"></a>Comments</h3><p>不能使用脚本控制<br>或者还是得调用MISO那一套</p>
<h2 id="SplicePlot"><a href="#SplicePlot" class="headerlink" title="SplicePlot"></a><a href="http://montgomerylab.stanford.edu/spliceplot/index.html" target="_blank" rel="noopener">SplicePlot</a></h2><h3 id="Installation-amp-Dependencies-2"><a href="#Installation-amp-Dependencies-2" class="headerlink" title="Installation &amp; Dependencies"></a>Installation &amp; Dependencies</h3><p><a href="https://github.com/wueric/SplicePlot" target="_blank" rel="noopener">GH page</a></p>
<ul>
<li><p>Python requirements:</p>
<ul>
<li>Python 2.6 or newer . If running Python 2.6, download and install the argparse module</li>
<li>numpy version 1.7.1 or newer </li>
<li>scipy version 0.11.0 or newer</li>
<li>matplotlib version 1.2.0 or newer</li>
<li>pandas version 0.11.0 or newer</li>
<li>pysam version 0.7 or newer</li>
</ul>
</li>
<li><p>Other requirements:</p>
<ul>
<li>SAMtools must be installed and be in your $PATH</li>
<li>Tabix must be installed and be in your $PATH</li>
</ul>
</li>
</ul>
<ol>
<li>Install the above dependencies.</li>
<li>Download the SplicePlot .zip and unpack the archive in the desired directory</li>
<li>Check if all of the necessary Python packages are available by running the command python check_module_availability.py in the SplicePlot directory.</li>
</ol>
<p>Cannot be found on conda cloud.</p>
<h3 id="Usage-amp-Materials-needed-2"><a href="#Usage-amp-Materials-needed-2" class="headerlink" title="Usage &amp; Materials needed"></a>Usage &amp; Materials needed</h3><!-- needs updates of summary -->
<p>Details are shown <a href="http://montgomerylab.stanford.edu/spliceplot/index.html#running-spliceplot" target="_blank" rel="noopener">here</a>.</p>
<p>Input files:</p>
<ul>
<li>A .vcf which has been sorted, bgzipped, and indexed using tabix.</li>
<li>A .gtf annotation file containing only exons which has been sorted by position, bgzipped, and then indexed using tabix.</li>
<li>.bam files containing the aligned RNA-seq data.</li>
<li>A mapping file which assigns the individual IDs from the .vcf to file paths for the .bam files as shown below.<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Individual_ID /path/to/bam/file.bam</span><br><span class="line">Another_ID /path/to/other/bam/file.bam</span><br></pre></td></tr></table></figure>
</li>
</ul>
<h3 id="Output-example-2"><a href="#Output-example-2" class="headerlink" title="Output example"></a>Output example</h3><p><img src="http://montgomerylab.stanford.edu/spliceplot/_images/sashimi_sample.png" alt="SplicePlot sashimi_plot"></p>
<p>Noted that the exon connections are presented as <strong>splicing ratio</strong>, which is calulated as:<br><img src="http://montgomerylab.stanford.edu/spliceplot/_images/math/453a31526e1d09c7ec4c25bc12199f8824f1facb.png" alt="SplicePlot splicing_ratio"></p>
<p>And the package can also print out <a href="http://montgomerylab.stanford.edu/spliceplot/index.html#hive-plots" target="_blank" rel="noopener">hive-plots</a> and <a href="http://montgomerylab.stanford.edu/spliceplot/index.html#structure-plots" target="_blank" rel="noopener">Structure-plots</a></p>
<h3 id="Comments-2"><a href="#Comments-2" class="headerlink" title="Comments"></a>Comments</h3><p>The updates stop at 2013.<br>The package was built under python 2.x</p>
<h1 id="Exploring"><a href="#Exploring" class="headerlink" title="Exploring"></a>Exploring</h1><h2 id="Create-a-conda-environment-for-sashimi"><a href="#Create-a-conda-environment-for-sashimi" class="headerlink" title="Create a conda environment for sashimi"></a>Create a conda environment for sashimi</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">conda create -n sashimi python=2.7</span><br><span class="line">conda install -c bioconda misopy</span><br><span class="line"># find a lovely path</span><br><span class="line">git clone https://github.com/wueric/SplicePlot.git</span><br></pre></td></tr></table></figure>
<h2 id="Splice-Plot"><a href="#Splice-Plot" class="headerlink" title="Splice Plot"></a>Splice Plot</h2><h3 id="Test"><a href="#Test" class="headerlink" title="Test"></a>Test</h3><blockquote>
<p>To test SplicePlot, run the command <code>python plot.py test_files/test.p pickle settings_file</code> in the SplicePlot directory.</p>
</blockquote>
<p>Something’s wrong…<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#&gt; Reading settings from settings_file...</span><br><span class="line">#&gt; Parsing settings for hive plot...</span><br><span class="line">#&gt; Parsing settings for structure plot...</span><br><span class="line">#&gt; Parsing sashimi plot settings...</span><br><span class="line">#&gt; Done reading settings.</span><br><span class="line">#&gt; Traceback (most recent call last):</span><br><span class="line">#&gt;   File &quot;plot.py&quot;, line 111, in &lt;module&gt;</span><br><span class="line">#&gt;     data_frame = pickle.load(pickle_file)</span><br><span class="line">#&gt; TypeError: _reconstruct: First argument must be a sub-type of ndarray</span><br></pre></td></tr></table></figure></p>
<p>Cannot find the pickle file.</p>
<p>说起来真是奇怪为什么一定要写SNP…写junction site不就完了。。<br>不太懂这个工具的逻辑</p>
<h2 id="MISO-sashimi-plot"><a href="#MISO-sashimi-plot" class="headerlink" title="MISO: sashimi_plot"></a>MISO: sashimi_plot</h2><p>find MISO path in conda env <code>~/anaconda3/envs/sashimi/lib/python2.7/site-packages/misopy/</code></p>
<h3 id="Test-1"><a href="#Test-1" class="headerlink" title="Test"></a>Test</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">cd sashimi_plot/</span><br><span class="line">index_gff --index test-data/events.gff test-data/event-data/</span><br><span class="line">#&gt; vast amounts of outputs</span><br><span class="line">#&gt; then copy the setting.txt on the website to your path/settings</span><br><span class="line">sashimi_plot --plot-event &quot;chr17:45816186:45816265:-@chr17:45815912:45815950:-@chr17:45814875:45814965:-&quot; test-data/event-data/ settings/sashimi_plot_settings.txt --output-dir test-plot</span><br></pre></td></tr></table></figure>
<p>Test complete. No errors found.</p>
<h1 id="在自己的数据上进行测试"><a href="#在自己的数据上进行测试" class="headerlink" title="在自己的数据上进行测试"></a>在自己的数据上进行测试</h1><ul>
<li>prepare bams</li>
<li>prepare gff/gtf</li>
<li>prepare setting file</li>
<li>create a wrapper (currently in R)</li>
</ul>
<h2 id="Prepare-the-alternative-isoforms-annotations-for-MISO-sashimi-plot"><a href="#Prepare-the-alternative-isoforms-annotations-for-MISO-sashimi-plot" class="headerlink" title="Prepare the alternative isoforms annotations for MISO::sashimi_plot"></a>Prepare the alternative isoforms annotations for MISO::sashimi_plot</h2><p>gencode.v27.gtf is fine.</p>
<h2 id="Preparing-GFF3"><a href="#Preparing-GFF3" class="headerlink" title="Preparing GFF3"></a>Preparing GFF3</h2><p>use <a href="http://genes.mit.edu/burgelab/miso/scripts/gtf2gff3.pl" target="_blank" rel="noopener">this script</a> to convert GTFs to GFF3, which needs a <code>.cfg</code> files<br>I found <a href="https://github.com/bioperl/Bio-FeatureIO/blob/master/t/data/gtf2gff3.cfg" target="_blank" rel="noopener">this template</a> online.</p>
<p>GFF3转换好像需要什么trick…<br>网上下的gtf2gff3.pl不太好使</p>
<p>于是寄几改了一个别人的gtf2gff3.py脚本<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/python</span><br><span class="line">&quot;&quot;&quot;</span><br><span class="line">Convert ensemble gtf to gff3 for MISO:sashimi_plot</span><br><span class="line">Copyright (C)</span><br><span class="line">__Author__: Jason Li</span><br><span class="line">&quot;&quot;&quot;</span><br><span class="line"></span><br><span class="line">import sys</span><br><span class="line">import re</span><br><span class="line"></span><br><span class="line">chr = &apos;chr&apos;</span><br><span class="line">keep_features = [&quot;CDS&quot;, &quot;UTR&quot;, &quot;exon&quot;, &quot;gene&quot;, &quot;transcript&quot;]</span><br><span class="line"></span><br><span class="line">if len(sys.argv) &lt; 2:</span><br><span class="line">  print(&quot;Run:\n\tpython &quot; + sys.argv[0] + &quot; input.gtf output.gff3&quot;)</span><br><span class="line">  exit()</span><br><span class="line"></span><br><span class="line">gtf_file = sys.argv[1]</span><br><span class="line">if len(sys.argv) == 2:</span><br><span class="line">  gff_file = gtf_file.rstrip(&apos;gtf&apos;) + &apos;gff3&apos;</span><br><span class="line">else:</span><br><span class="line">  gff_file = sys.argv[2]</span><br><span class="line"></span><br><span class="line">### Parse gtf file</span><br><span class="line">gtf = open(gtf_file, &apos;r&apos;)</span><br><span class="line">gff = open(gff_file, &apos;w&apos;)</span><br><span class="line"></span><br><span class="line">gtf_lines = gtf.readlines()</span><br><span class="line">line_num = 0</span><br><span class="line">for line in gtf_lines:</span><br><span class="line">  line_num += 1</span><br><span class="line">  if re.match(r&apos;^#&apos;, line):</span><br><span class="line">    continue</span><br><span class="line">  featureline = line.strip().split(&apos;\t&apos;)</span><br><span class="line">  [seq, source, feature, start, end, score, strand, frame, attributes] = featureline</span><br><span class="line">  if feature not in keep_features:</span><br><span class="line">    continue</span><br><span class="line"></span><br><span class="line">  seq = seq</span><br><span class="line">  if feature == &apos;transcript&apos;:</span><br><span class="line">    feature = &apos;mRNA&apos;</span><br><span class="line"></span><br><span class="line">  attributes_array = attributes.replace(&apos;&quot;&apos;,&apos;&apos;).strip(&apos;;&apos;).split(&apos;; &apos;)</span><br><span class="line">  attr_dict = dict(map(lambda x: x.split(), attributes_array))</span><br><span class="line">  </span><br><span class="line">  if feature == &apos;gene&apos;:</span><br><span class="line">    # if I have time, this should be able to be controled by a config file</span><br><span class="line">    geneID = attr_dict[&apos;gene_name&apos;]</span><br><span class="line">    Name = geneID</span><br><span class="line">    gff.write(</span><br><span class="line">      &apos;\t&apos;.join([seq, source, feature, start, end, score, strand, frame]) + &apos;\t&apos; + </span><br><span class="line">      &apos;ID=&apos; + geneID + &apos;;Name=&apos;+Name +&apos;\n&apos;)</span><br><span class="line">  elif feature == &apos;mRNA&apos;:</span><br><span class="line">    #count = 0</span><br><span class="line">    Parent = geneID</span><br><span class="line">    transID = geneID + &apos;.&apos; + attr_dict[&apos;transcript_name&apos;]</span><br><span class="line">    gff.write(</span><br><span class="line">      &apos;\t&apos;.join([seq, source, feature, start, end, score, strand, frame]) + &apos;\t&apos; + </span><br><span class="line">      &apos;ID=&apos; + transID + &apos;;Parent=&apos;+ Parent +&apos;\n&apos;)</span><br><span class="line">  elif feature == &apos;exon&apos;:</span><br><span class="line">    #count += 1</span><br><span class="line">    Parent = transID</span><br><span class="line">    exonID = transID + &apos;.&apos; + attr_dict[&apos;exon_id&apos;]</span><br><span class="line">    gff.write(</span><br><span class="line">      &apos;\t&apos;.join([seq, source, feature, start, end, score, strand, frame]) + &apos;\t&apos; + </span><br><span class="line">      &apos;ID=&apos; + exonID + &apos;;Parent=&apos;+ Parent +&apos;\n&apos;)</span><br><span class="line">  else:</span><br><span class="line">    pass</span><br><span class="line"></span><br><span class="line">  if line_num % 10000 == 0:</span><br><span class="line">    print &quot;&#123;&#125; lines processed.&quot;.format(line_num)</span><br><span class="line">  # gff.write(</span><br><span class="line">  #   &apos;\t&apos;.join([seq, source, feature, start, end, score, strand, frame]) + &apos;\t&apos; + </span><br><span class="line">  #   &apos;Parent=&apos; + ID + &apos;\n&apos;)</span><br><span class="line"></span><br><span class="line">gff.close()</span><br><span class="line">gtf.close()</span><br></pre></td></tr></table></figure></p>
<p>使用index_gff对这个gff3进行index</p>
<p>报错，找不到某个pickle文件，懒得查错到底是什么，，明明别的都输出的好好的。<br>在<code>~/anaconda3/envs/sashimi/lib/python2.7/site-packages/misopy/pickle_utils.py</code>的写文件函数里加了个<code>try...except</code>结构<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">def write_pickled_file(obj_to_pickle, pickled_filename):</span><br><span class="line">    try: </span><br><span class="line">        pickled_file = file(pickled_filename, &apos;w&apos;)</span><br><span class="line">        pickle.dump(obj_to_pickle, pickled_file, -1)</span><br><span class="line">        #del pickled_file</span><br><span class="line">        pickled_file.close()</span><br><span class="line">    except Exception,e:</span><br><span class="line">        print Exception,&quot;:&quot;,e</span><br></pre></td></tr></table></figure></p>
<h2 id="Prepare-setting-file"><a href="#Prepare-setting-file" class="headerlink" title="Prepare setting file"></a>Prepare setting file</h2><p>经过测试，miso相关的文件完全不需要也可以进行sashimi plot<br>仅仅需要一些plot相关的参数以及bam的路径以及对应各个样本的颜色信息和总reads信息(for RPKM calculation)</p>
<p>通过测试的minimum setting file looks like this:<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">[plotting]</span><br><span class="line"># Dimensions of figure to be plotted (in inches)</span><br><span class="line">fig_width = 7</span><br><span class="line">fig_height = 5</span><br><span class="line"># Factor to scale down introns and exons by</span><br><span class="line">intron_scale = 30</span><br><span class="line">exon_scale = 4</span><br><span class="line"># Whether to use a log scale or not when plotting</span><br><span class="line">logged = True</span><br><span class="line">font_size = 6</span><br><span class="line"> </span><br><span class="line"># Max y-axis</span><br><span class="line"># ymax = 150</span><br><span class="line"> </span><br><span class="line"># Whether to plot posterior distributions inferred by MISO</span><br><span class="line">show_posteriors = False</span><br><span class="line"> </span><br><span class="line"># Whether to show posterior distributions as bar summaries</span><br><span class="line">bar_posteriors = False</span><br><span class="line"> </span><br><span class="line"># Whether to plot the number of reads in each junction</span><br><span class="line">number_junctions = True</span><br><span class="line"> </span><br><span class="line">resolution = .5</span><br><span class="line">posterior_bins = 40</span><br><span class="line">gene_posterior_ratio = 5</span><br><span class="line"> </span><br><span class="line"> </span><br><span class="line">[data]</span><br><span class="line">bam_prefix = ./bams/</span><br><span class="line"> </span><br><span class="line">bam_files = [</span><br><span class="line">    &quot;heartWT1.sorted.bam&quot;,</span><br><span class="line">    &quot;heartWT2.sorted.bam&quot;,</span><br><span class="line">    &quot;heartKOa.sorted.bam&quot;,</span><br><span class="line">    &quot;heartKOb.sorted.bam&quot;]</span><br><span class="line"> </span><br><span class="line"># List of colors for read denisites of each sample</span><br><span class="line">colors = [</span><br><span class="line">    &quot;#CC0011&quot;,</span><br><span class="line">    &quot;#CC0011&quot;,</span><br><span class="line">    &quot;#FF8800&quot;,</span><br><span class="line">    &quot;#FF8800&quot;]</span><br><span class="line"> </span><br><span class="line"># Number of mapped reads in each sample</span><br><span class="line"># (Used to normalize the read density for RPKM calculation)</span><br><span class="line">coverages = [</span><br><span class="line">    6830944,</span><br><span class="line">    14039751,</span><br><span class="line">    4449737,</span><br><span class="line">    6720151]</span><br></pre></td></tr></table></figure></p>
<h2 id="Preparing-bams-bais"><a href="#Preparing-bams-bais" class="headerlink" title="Preparing bams/bais"></a>Preparing bams/bais</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">ls $PWD/01.STAR/*/*/*.bam | awk -F &quot;/&quot; &apos;&#123;print $8&quot;\t&quot;$9&quot;\t&quot;$0&#125;&apos; &gt; /path/to/data/sashimi_plot/bams.list</span><br><span class="line"></span><br><span class="line">source activate sashimi</span><br><span class="line"># 好的十分暴力</span><br><span class="line">cat bams.list | awk &apos;&#123;print &quot;samtools index &quot;$3&#125;&apos; | bash</span><br><span class="line"># 等不了一秒星人必须要监视一下才能按捺住寄几, 于是在另一个窗口</span><br><span class="line">b=`ls 01.STAR/*/*/*.bai | wc -l`</span><br><span class="line">sleep 360</span><br><span class="line">c=`ls 01.STAR/*/*/*.bai | wc -l`</span><br><span class="line">until [[ b -eq c ]]; do b=$c; sleep 360; c=`ls 01.STAR/*/*/*.bai | wc -l` ; done; serverChan -t &quot;bai_produce_done&quot;</span><br></pre></td></tr></table></figure>
<p>妈的真是慢啊。。但我又懒得写多核控制shell..啊不其实是不会。。</p>
<p>=== update ===</p>
<h1 id="由于登陆节点真的太慢了所以直接投计算节点了，超快！"><a href="#由于登陆节点真的太慢了所以直接投计算节点了，超快！" class="headerlink" title="由于登陆节点真的太慢了所以直接投计算节点了，超快！"></a>由于登陆节点真的太慢了所以直接投计算节点了，超快！</h1><blockquote>
<p>版权声明： 本博客所有文章除特别声明外，均采用<a href="https://creativecommons.org/licenses/by-nc-sa/3.0/cn/deed.zh" target="_blank" rel="noopener">CC BY-NC-SA 3.0 CN</a>许可协议。转载请注明出处！</p>
</blockquote>

      
    </div>

    <div>
      
        

      
    </div>

    <div>
      
        

      
    </div>

    <div>
      
        

      
    </div>

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
          
            <a href="/tags/Tools/" rel="tag"># Tools</a>
          
        </div>
      

      
        
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/05/07/isoform-function-in-literature/" rel="next" title="Systematic evaluation of isoform function in literature reports of alternative splicing">
                <i class="fa fa-chevron-left"></i> Systematic evaluation of isoform function in literature reports of alternative splicing
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/06/16/dplyr-mutate-new-column-based-on-multiple-selected-columns/" rel="prev" title="dplyr: mutate new column based on multiple selected columns">
                dplyr: mutate new column based on multiple selected columns <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
      <div id="lv-container" data-id="city" data-uid="MTAyMC8yODY0NC81MjE1"></div>
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/images/avatar.jpg"
               alt="Jason Li" />
          <p class="site-author-name" itemprop="name">Jason Li</p>
           
              <p class="site-description motion-element" itemprop="description">Notes about R & python</p>
          
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">24</span>
                <span class="site-state-item-name">日志</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">4</span>
                <span class="site-state-item-name">分类</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">34</span>
                <span class="site-state-item-name">标签</span>
              </a>
            </div>
          

        </nav>

        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="https://github.com/lijxug" target="_blank" title="GitHub">
                  
                    <i class="fa fa-fw fa-github"></i>
                  
                  GitHub
                </a>
              </span>
            
          
        </div>

        
        

        
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#Abstract"><span class="nav-number">1.</span> <span class="nav-text">Abstract</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#What-is-sashimi-plot"><span class="nav-number">2.</span> <span class="nav-text">What is sashimi_plot?</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#现存的几种方案"><span class="nav-number">3.</span> <span class="nav-text">现存的几种方案</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#MISO"><span class="nav-number">3.1.</span> <span class="nav-text">MISO</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Installation-amp-Dependencies"><span class="nav-number">3.1.1.</span> <span class="nav-text">Installation &amp; Dependencies</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Usage-amp-Materials-needed"><span class="nav-number">3.1.2.</span> <span class="nav-text">Usage &amp; Materials needed</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Output-example"><span class="nav-number">3.1.3.</span> <span class="nav-text">Output example</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Comments"><span class="nav-number">3.1.4.</span> <span class="nav-text">Comments</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#IGV"><span class="nav-number">3.2.</span> <span class="nav-text">IGV</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Installation-amp-Dependencies-1"><span class="nav-number">3.2.1.</span> <span class="nav-text">Installation &amp; Dependencies</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Usage-amp-Materials-needed-1"><span class="nav-number">3.2.2.</span> <span class="nav-text">Usage &amp; Materials needed</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Output-example-1"><span class="nav-number">3.2.3.</span> <span class="nav-text">Output example</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Comments-1"><span class="nav-number">3.2.4.</span> <span class="nav-text">Comments</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#SplicePlot"><span class="nav-number">3.3.</span> <span class="nav-text">SplicePlot</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Installation-amp-Dependencies-2"><span class="nav-number">3.3.1.</span> <span class="nav-text">Installation &amp; Dependencies</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Usage-amp-Materials-needed-2"><span class="nav-number">3.3.2.</span> <span class="nav-text">Usage &amp; Materials needed</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Output-example-2"><span class="nav-number">3.3.3.</span> <span class="nav-text">Output example</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Comments-2"><span class="nav-number">3.3.4.</span> <span class="nav-text">Comments</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Exploring"><span class="nav-number">4.</span> <span class="nav-text">Exploring</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#Create-a-conda-environment-for-sashimi"><span class="nav-number">4.1.</span> <span class="nav-text">Create a conda environment for sashimi</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Splice-Plot"><span class="nav-number">4.2.</span> <span class="nav-text">Splice Plot</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Test"><span class="nav-number">4.2.1.</span> <span class="nav-text">Test</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#MISO-sashimi-plot"><span class="nav-number">4.3.</span> <span class="nav-text">MISO: sashimi_plot</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Test-1"><span class="nav-number">4.3.1.</span> <span class="nav-text">Test</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#在自己的数据上进行测试"><span class="nav-number">5.</span> <span class="nav-text">在自己的数据上进行测试</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#Prepare-the-alternative-isoforms-annotations-for-MISO-sashimi-plot"><span class="nav-number">5.1.</span> <span class="nav-text">Prepare the alternative isoforms annotations for MISO::sashimi_plot</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Preparing-GFF3"><span class="nav-number">5.2.</span> <span class="nav-text">Preparing GFF3</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Prepare-setting-file"><span class="nav-number">5.3.</span> <span class="nav-text">Prepare setting file</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Preparing-bams-bais"><span class="nav-number">5.4.</span> <span class="nav-text">Preparing bams/bais</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#由于登陆节点真的太慢了所以直接投计算节点了，超快！"><span class="nav-number">6.</span> <span class="nav-text">由于登陆节点真的太慢了所以直接投计算节点了，超快！</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy;  2017 - 
  <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Jason Li</span>
</div>


<div class="powered-by">
  由 <a class="theme-link" href="https://hexo.io">Hexo</a> 强力驱动
</div>

<div class="theme-info">
  主题 -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next">
    NexT.Pisces
  </a>
</div>


        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
      
    </span>
  

  
    <span class="site-pv">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
      
    </span>
  
</div>


        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  


  











  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>

  
  <script type="text/javascript" src="/lib/canvas-nest/canvas-nest.min.js"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.1"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.1"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.1"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.1"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.1"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.1"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.1"></script>



  


  




	





  





  





  
    <script type="text/javascript">
      (function(d, s) {
        var j, e = d.getElementsByTagName(s)[0];
        if (typeof LivereTower === 'function') { return; }
        j = d.createElement(s);
        j.src = 'https://cdn-city.livere.com/js/embed.dist.js';
        j.async = true;
        e.parentNode.insertBefore(j, e);
      })(document, 'script');
    </script>
  






  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (search_path.endsWith("json")) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
        processEscapes: true,
        skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
      },
      TeX: {equationNumbers: { autoNumber: "AMS" }}
    });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
  


  

  

</body>
</html>
